dag_in_context/src/pretty_print.rs

```diff
@@ -0,0 +1,831 @@
+            if !pp.table.contains_key(fresh_var) {
+                pp.log.push(fresh_var.into());
+                pp.table.insert(fresh_var.into(), SymbolLog::Expr(new_expr));
+            }
+            Expr::Symbolic(fresh_var.into())
+        };
+
+        let num_shared = Rc::strong_count(expr);
+        let fold_or_plain = |pp: &mut PrettyPrinter, new_expr: Expr| {
+            let size = &new_expr
+                .to_string()
+                .replace(&['(', ')', ' ', ','][..], "") //don't count those char when computing size
+                .len();
+            if fold_when(num_shared, *size) {
+                fold(pp, new_expr)
+            } else {
+                new_expr
+            }
+        };
+
+        fn handle_assum(
+            assum: &Assumption,
+            pp: &mut PrettyPrinter,
+            fold_when: &dyn Fn(usize, usize) -> bool,
+            to_rust: bool,
+        ) -> String {
+            let old_assume_binding = pp
+                .symbols
+                .get(&Symbols::Assumption(assum.to_ref()))
+                .unwrap()
+                .clone();
+            if !pp.table.contains_key(&old_assume_binding) {
+                let new_assum = match assum {
+                    Assumption::InFunc(_) => assum.clone(),
+                    Assumption::InIf(cond, left, right) => {
+                        let left = pp.fold_expr(left, fold_when, to_rust);
+                        let right = pp.fold_expr(right, fold_when, to_rust);
+                        Assumption::InIf(*cond, Rc::new(left), Rc::new(right))
+                    }
+                    Assumption::InLoop(inputs, body) => {
+                        let inputs = pp.fold_expr(inputs, fold_when, to_rust);
+                        let body = pp.fold_expr(body, fold_when, to_rust);
+                        Assumption::InLoop(Rc::new(inputs), Rc::new(body))
+                    }
+                    Assumption::InSwitch(cond, inputs, branch) => {
+                        let inputs = pp.fold_expr(inputs, fold_when, to_rust);
+                        let branch = pp.fold_expr(branch, fold_when, to_rust);
+                        Assumption::InSwitch(*cond, Rc::new(inputs), Rc::new(branch))
+                    }
+                    Assumption::WildCard(_) => assum.clone(),
+                };
+                pp.log.push(old_assume_binding.clone());
+                pp.table
+                    .insert(old_assume_binding.clone(), SymbolLog::Assumption(new_assum));
+            }
+
+            old_assume_binding
+        }
+
+        fn handle_type(ty: &Type, pp: &mut PrettyPrinter) -> Type {
+            let ty_str = pp.symbols.get(&Symbols::Type(ty.clone())).unwrap().clone();
+            if !pp.table.contains_key(&ty_str) {
+                pp.log.push(ty_str.clone());
+                pp.table.insert(ty_str.clone(), SymbolLog::Type(ty.clone()));
+            }
+            Type::Symbolic(ty_str)
+        }
+
+        match expr.as_ref() {
```

Comment:
Can we use some new methods `map_children`, `map_types`, and `map_assumptions` to make this much simpler?